File: elixir/lib/solutions/0238_product_of_array_except_self/product_of_array_except_self.ex
defmodule LeetCode.Solutions.ProductOfArrayExceptSelf do
  @moduledoc false

  @spec call(nums :: [integer]) :: [integer]
  def call(nums) do
    # This solution doesn't work because it takes too long
    # to execute all of the test cases in LeetCode.
    # See my notes on this problem for more information.

    {prefix_products, _} =
      Enum.reduce(nums, {[], 1}, fn num, {products, prefix} ->
        {[prefix | products], prefix * num}
      end)

    nums
    |> Enum.reverse()
    |> Enum.with_index()
    |> Enum.reduce({prefix_products, 1}, fn {num, index}, {products, suffix} ->
      {List.update_at(products, index, &(&1 * suffix)), suffix * num}
    end)
    |> elem(0)
    |> Enum.reverse()
  end
end

File: ruby/lib/solutions/0238_product_of_array_except_self/product_of_array_except_self.rb
# frozen_string_literal: true

def product_except_self(nums)
  prefix = 1
  suffix = 1
  product = []

  nums.each do |num|
    product << prefix
    prefix *= num
  end

  (nums.length - 1).downto(0) do |index|
    product[index] *= suffix
    suffix *= nums[index]
  end

  product
end
